### PR TITLE
PHP 7.3 compatibility

### PR DESCRIPTION
**Pull request by Jan-E**
@remicollet This was the PR I had been testing.

Ref: https://externals.io/message/102685

### CODE REVIEW

**Comment by remicollet** [comment]
Test suite OK with 7.2

With 7.3 we have a segfault

```
(gdb) bt
#0  zend_mm_alloc_small (bin_num=4, size=40, heap=0x7ffff3600040) at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/Zend/zend_alloc.c:1285
#1  _emalloc_40 () at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/Zend/zend_alloc.c:2419
#2  0x00007fffec38bebb in zend_string_alloc (persistent=0, len=10) at /opt/remi/php73/root/usr/include/php/Zend/zend_string.h:133
#3  zend_string_init (persistent=0, len=10, str=0x7fffec3908d5 "content-id") at /opt/remi/php73/root/usr/include/php/Zend/zend_string.h:155
#4  mailparse_get_part_data (part=0x7ffff3668680, return_value=return_value@entry=0x7ffff361e850) at /home/php/git/mailparse/mailparse.c:1479
#5  0x00007fffec38c513 in zif_mailparse_msg_get_part_data (execute_data=<optimized out>, return_value=0x7ffff361e850) at /home/php/git/mailparse/mailparse.c:1513
#6  0x0000555555874e7a in ZEND_DO_ICALL_SPEC_RETVAL_USED_HANDLER () at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/Zend/zend_vm_execute.h:690
#7  execute_ex (ex=0x7ffff3600040) at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/Zend/zend_vm_execute.h:54541
#8  0x000055555587a023 in zend_execute (op_array=op_array@entry=0x7ffff36862a0, return_value=0x0, return_value@entry=0x7ffff36a0940)
    at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/Zend/zend_vm_execute.h:59965
#9  0x00005555557e9b12 in zend_execute_scripts (type=type@entry=8, retval=0x7ffff36a0940, retval@entry=0x0, file_count=-211689424, file_count@entry=3)
    at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/Zend/zend.c:1564
#10 0x0000555555789a50 in php_execute_script (primary_file=0x7fffffffc440) at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/main/main.c:2608
#11 0x000055555587c47c in do_cli (argc=64, argv=0x555555bf8440) at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/sapi/cli/php_cli.c:1002
#12 0x000055555564118b in main (argc=64, argv=0x555555bf8440) at /usr/src/debug/php73-php-7.3.0~alpha3-1.fc27.remi.x86_64/sapi/cli/php_cli.c:1395

```

**Comment by cmb69** [comment]
On a quick glance, this doesn't look like a mailparse issue. After all, what could be wrong with [this zend_string_init](https://github.com/Jan-E/pecl-mail-mailparse/blob/05e2c643281f669961f87df324f52d38ff544b21/mailparse.c#L1479)? Stack frame #3 also looks okay.

BTW: why do you attempt to achieve PHP 7.3 compatibility, for a [PHP 4 extension](https://github.com/php/pecl-mail-mailparse/blob/master/README#L1)? ;)

File: mailparse.c [hunk 1]
```diff
@@ -978,12 +978,21 @@ PHP_FUNCTION(mailparse_stream_encode)
 	buf = emalloc(bufsize);
 	RETVAL_TRUE;
 
+#if PHP_VERSION_ID >= 70300
+	conv = mbfl_convert_filter_new(mbfl_no2encoding(mbfl_no_encoding_8bit),
```
Comment (cmb69) [inline mailparse.c:982]: [According to @nikic](https://externals.io/message/102685#102692) this line should be:
````
	conv = mbfl_convert_filter_new(&mbfl_no_encoding_8bit,
````

**Comment by Jan-E** [comment]
> On a quick glance, this doesn't look like a mailparse issue. After all, what could be wrong with this zend_string_init? Stack frame #3 also looks okay.

I moved the zend-string_init to the top of the block. Now the segfaults starts at https://github.com/Jan-E/pecl-mail-mailparse/blob/c90c5fe18aa8b1a76a5269d36661eeede22c247e/mailparse.c#L1476 (!), near the same line as before (1479).

```
		if (part->boundary)
	6AB736A2  push        0Ch  
	6AB736A4  push        offset string "content-base" (6AB78ACCh)  
	6AB736A9  push        esi  
	6AB736AA  call        dword ptr [__imp__add_assoc_string_ex (6AB7816Ch)]  
>	6AB736B0  mov         eax,dword ptr [ebx+78h]  
	6AB736B3  add         esp,10h  
	6AB736B6  test        eax,eax  
	6AB736B8  je          mailparse_get_part_data+22Ch (6AB736CCh)  
			add_assoc_string(return_value, "content-boundary", part->boundary);

```
Backtrace:
```
 	[Inline Frame] php7.dll!zend_mm_alloc_small(_zend_mm_heap *) Line 1285	C
 	[Inline Frame] php7.dll!zend_mm_alloc_heap(_zend_mm_heap * heap, unsigned int) Line 1356	C
 	php7.dll!emalloc(unsigned int size) Line 2496	C
 	[Inline Frame] php7.dll!zend_string_alloc(unsigned int) Line 133	C
 	[Inline Frame] php7.dll!zend_string_init(const char * persistent, unsigned int) Line 155	C
 	[Inline Frame] php7.dll!_zend_hash_str_add_or_update_i(_zend_array *) Line 774	C
 	php7.dll!zend_hash_str_update(_zend_array * ht, const char * str, unsigned int len, _zval_struct * pData) Line 837	C
 	[Inline Frame] php7.dll!zend_symtable_str_update(_zend_array * ht, const char *) Line 531	C
 	php7.dll!add_assoc_string_ex(_zval_struct * arg, const char * key, unsigned int key_len, const char * str) Line 1419	C
>	php_mailparse.dll!mailparse_get_part_data(_php_mimepart * part, _zval_struct * return_value) Line 1476	C
 	php_mailparse.dll!zif_mailparse_msg_get_part_data(_zend_execute_data * execute_data, _zval_struct * return_value) Line 1512	C
 	php7.dll!ZEND_DO_ICALL_SPEC_RETVAL_USED_HANDLER(_zend_execute_data * execute_data) Line 701	C
 	php7.dll!execute_ex(_zend_execute_data * ex) Line 54410	C
 	php7.dll!zend_execute(_zend_op_array * op_array, _zval_struct * return_value) Line 59965	C
 	php7.dll!zend_execute_scripts(int type, _zval_struct * retval, int file_count, ...) Line 1564	C
 	php7.dll!php_execute_script(_zend_file_handle * primary_file) Line 2608	C
 	php.exe!do_cli(int argc, char * * argv) Line 1003	C
 	php.exe!main(int argc, char * * argv) Line 1397	C
```

**Comment by remicollet** [comment]
> On a quick glance, this doesn't look like a mailparse issue. After all, what could be wrong with this zend_string_init? 

@cmb69 this can be a memory corruption, raise by something wrong in some previous part (double-free, ...)

FYI the extension also have issues with 7.2 (https://bugs.php.net/bug.php?id=76561) and newer GCC

File: mailparse.c (hunk 1 above)
Comment (remicollet) [reply mailparse.c:982]: You mean &mbfl_encoding_8bit ?
The problem is that is NOT a public symbol in the bundled library, so is not usable `(modules/mailparse.so: undefined symbol: mbfl_encoding_8bit)`

**Comment by remicollet** [comment]
I have merged this PR, keep this open for discussion around the segfault

**Comment by remicollet** [comment]
Valgrind reports something interesting:

```
==26680== Invalid read of size 4
==26680==    at 0x3AD632: gc_check_possible_root (zend_gc.h:72)
==26680==    by 0x3AD632: i_zval_ptr_dtor (zend_variables.h:51)
==26680==    by 0x3AD632: zend_array_destroy (zend_hash.c:1472)
==26680==    by 0x3DCF47: zend_assign_to_variable (zend_execute.h:110)
==26680==    by 0x3DCF47: ZEND_ASSIGN_SPEC_CV_VAR_RETVAL_UNUSED_HANDLER (zend_vm_execute.h:45071)
==26680==    by 0x4258EC: execute_ex (zend_vm_execute.h:59325)
==26680==    by 0x42E022: zend_execute (zend_vm_execute.h:59965)
==26680==    by 0x39DB11: zend_execute_scripts (zend.c:1564)
==26680==    by 0x33DA4F: php_execute_script (main.c:2608)
==26680==    by 0x43047B: do_cli (php_cli.c:1002)
==26680==    by 0x1F518A: main (php_cli.c:1395)
==26680==  Address 0x10a90d64 is 4 bytes inside a block of size 40 free'd
==26680==    at 0x4C2DD18: free (vg_replace_malloc.c:530)
==26680==    by 0x3AD7C4: i_zval_ptr_dtor (zend_variables.h:49)
==26680==    by 0x3AD7C4: zend_array_destroy (zend_hash.c:1472)
==26680==    by 0x10719F79: php_mimepart_free (php_mailparse_mime.c:356)
==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
==26680==    by 0x10719E62: php_mimepart_free (php_mailparse_mime.c:333)
==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
==26680==    by 0x10719E62: php_mimepart_free (php_mailparse_mime.c:333)
==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
==26680==    by 0x10719E62: php_mimepart_free (php_mailparse_mime.c:333)
==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
==26680==    by 0x10719E62: php_mimepart_free (php_mailparse_mime.c:333)
==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
==26680==  Block was alloc'd at
==26680==    at 0x4C2CB6B: malloc (vg_replace_malloc.c:299)
==26680==    by 0x3737D8: __zend_malloc (zend_alloc.c:2898)
==26680==    by 0x39FA11: UnknownInlinedFun (zend_string.h:133)
==26680==    by 0x39FA11: UnknownInlinedFun (zend_string.h:155)
==26680==    by 0x39FA11: add_assoc_string_ex (zend_API.c:1418)
==26680==    by 0x1071A5D4: php_mimepart_process_header (php_mailparse_mime.c:455)
==26680==    by 0x1071AB69: php_mimepart_process_line (php_mailparse_mime.c:635)
==26680==    by 0x1071AB69: php_mimepart_parse (php_mailparse_mime.c:724)
==26680==    by 0x10716184: zif_mailparse_msg_parse (mailparse.c:1059)
==26680==    by 0x4280B6: ZEND_DO_ICALL_SPEC_RETVAL_UNUSED_HANDLER (zend_vm_execute.h:645)
==26680==    by 0x4280B6: execute_ex (zend_vm_execute.h:54537)
==26680==    by 0x42E022: zend_execute (zend_vm_execute.h:59965)
==26680==    by 0x39DB11: zend_execute_scripts (zend.c:1564)
==26680==    by 0x33DA4F: php_execute_script (main.c:2608)
==26680==    by 0x43047B: do_cli (php_cli.c:1002)
==26680==    by 0x1F518A: main (php_cli.c:1395)

```

File: mailparse.c (hunk 1 above)
Comment (nikic) [reply mailparse.c:982]: I think the symbol is public, you just need an explicit include on `libmbfl/mbfl/mbfilter_8bit.h`.

But of course, your version works too, if performance is not an issue.

File: mailparse.c [hunk 2]
```diff
@@ -1159,9 +1168,13 @@ PHP_FUNCTION(mailparse_msg_get_structure)
 
 	mailparse_fetch_mimepart_resource(part, arg);
 
+#if PHP_VERSION_ID >= 70300
+	array_init(return_value);
+#else
```
Comment (nikic) [inline mailparse.c:1173]: You can drop the else branch here. array_init was already infallible before.

File: mailparse.c (hunk 1 above)
Comment (Jan-E) [reply mailparse.c:982]: Including ext/mbstring/libmbfl/mbfl/mbfilter_8bit.h fails with a 
`ext\mailparse\php_mailparse_mime.c(912): error C2101: '&' on constant`

Comment (nikic) [reply mailparse.c:982]: @Jan-E Are you using `&mbfl_no_encoding_8bit` or `&mbfl_encoding_8bit` when you get that error?

Comment (Jan-E) [reply mailparse.c:982]: &mbfl_encoding_8bit

**Comment by rlerdorf** [comment]
> Valgrind reports something interesting:
> 
> ```
> ==26680== Invalid read of size 4
> ==26680==    at 0x3AD632: gc_check_possible_root (zend_gc.h:72)
> ==26680==    by 0x3AD632: i_zval_ptr_dtor (zend_variables.h:51)
> ==26680==    by 0x3AD632: zend_array_destroy (zend_hash.c:1472)
> ==26680==    by 0x3DCF47: zend_assign_to_variable (zend_execute.h:110)
> ==26680==    by 0x3DCF47: ZEND_ASSIGN_SPEC_CV_VAR_RETVAL_UNUSED_HANDLER (zend_vm_execute.h:45071)
> ==26680==    by 0x4258EC: execute_ex (zend_vm_execute.h:59325)
> ==26680==    by 0x42E022: zend_execute (zend_vm_execute.h:59965)
> ==26680==    by 0x39DB11: zend_execute_scripts (zend.c:1564)
> ==26680==    by 0x33DA4F: php_execute_script (main.c:2608)
> ==26680==    by 0x43047B: do_cli (php_cli.c:1002)
> ==26680==    by 0x1F518A: main (php_cli.c:1395)
> ==26680==  Address 0x10a90d64 is 4 bytes inside a block of size 40 free'd
> ==26680==    at 0x4C2DD18: free (vg_replace_malloc.c:530)
> ==26680==    by 0x3AD7C4: i_zval_ptr_dtor (zend_variables.h:49)
> ==26680==    by 0x3AD7C4: zend_array_destroy (zend_hash.c:1472)
> ==26680==    by 0x10719F79: php_mimepart_free (php_mailparse_mime.c:356)
> ==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
> ==26680==    by 0x10719E62: php_mimepart_free (php_mailparse_mime.c:333)
> ==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
> ==26680==    by 0x10719E62: php_mimepart_free (php_mailparse_mime.c:333)
> ==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
> ==26680==    by 0x10719E62: php_mimepart_free (php_mailparse_mime.c:333)
> ==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
> ==26680==    by 0x10719E62: php_mimepart_free (php_mailparse_mime.c:333)
> ==26680==    by 0x3AD544: zend_hash_destroy (zend_hash.c:1398)
> ==26680==  Block was alloc'd at
> ==26680==    at 0x4C2CB6B: malloc (vg_replace_malloc.c:299)
> ==26680==    by 0x3737D8: __zend_malloc (zend_alloc.c:2898)
> ==26680==    by 0x39FA11: UnknownInlinedFun (zend_string.h:133)
> ==26680==    by 0x39FA11: UnknownInlinedFun (zend_string.h:155)
> ==26680==    by 0x39FA11: add_assoc_string_ex (zend_API.c:1418)
> ==26680==    by 0x1071A5D4: php_mimepart_process_header (php_mailparse_mime.c:455)
> ==26680==    by 0x1071AB69: php_mimepart_process_line (php_mailparse_mime.c:635)
> ==26680==    by 0x1071AB69: php_mimepart_parse (php_mailparse_mime.c:724)
> ==26680==    by 0x10716184: zif_mailparse_msg_parse (mailparse.c:1059)
> ==26680==    by 0x4280B6: ZEND_DO_ICALL_SPEC_RETVAL_UNUSED_HANDLER (zend_vm_execute.h:645)
> ==26680==    by 0x4280B6: execute_ex (zend_vm_execute.h:54537)
> ==26680==    by 0x42E022: zend_execute (zend_vm_execute.h:59965)
> ==26680==    by 0x39DB11: zend_execute_scripts (zend.c:1564)
> ==26680==    by 0x33DA4F: php_execute_script (main.c:2608)
> ==26680==    by 0x43047B: do_cli (php_cli.c:1002)
> ==26680==    by 0x1F518A: main (php_cli.c:1395)
> ```

This is still happening and it isn't limited to PHP 7.3. I can reproduce it against 7.1 and 7.2 as well. To reproduce, just run `make test`. One test should fail with a segfault. Then you can valgrind `php tests/parse_test_messages.php`

**Comment by DRoet** [comment]
I know this PR has already been merged, but it would be nice to have a release aswell now that `7.3` it out.

**Comment by remicollet** [comment]
@DRoet no, not ready, still broken with 7.2 and 7.3

**Comment by DRoet** [comment]
@remicollet fair enough, thanks for the status update. I just think more people will run into this since it fails on install @ `7.3`, but I agree the segfault is just as important ;)

**Comment by zbenc** [comment]
@nikic Any chance you could help find and fix the segfault bug caused by accessing members of the mailparse_get_part_data return value with PHP 7.3? Your help would be very much appreciated, thanks.

**Comment by cmb69** [comment]
The following patch seems to make valgrind happy:
````patch
 mailparse.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/mailparse.c b/mailparse.c
index df321c1..92b247c 100644
--- a/mailparse.c
+++ b/mailparse.c
@@ -1417,7 +1417,6 @@ static void add_header_reference_to_zval(char *headerkey, zval *return_value, zv
 	hash_key = zend_string_init(headerkey, strlen(headerkey), 0);
 	if ((headerval = zend_hash_find(Z_ARRVAL_P(headers), hash_key)) != NULL) {
 		ZVAL_DUP(&newhdr, headerval);
-		Z_SET_REFCOUNT_P(&newhdr, 1);
 		add_assoc_zval(return_value, headerkey, &newhdr);
 	}
 	zend_string_release(hash_key);
@@ -1434,7 +1433,7 @@ static int mailparse_get_part_data(php_mimepart *part, zval *return_value)
 	array_init(return_value);
 
 	/* get headers for this section */
-	ZVAL_DUP(&headers, &part->headerhash);
+	ZVAL_COPY(&headers, &part->headerhash);
 
 	add_assoc_zval(return_value, "headers", &headers);
 ````

PS: Even with this patch applied, 007.phpt still causes an OOB in https://github.com/php/pecl-mail-mailparse/blob/d6281bc9a6009ef66e2ea72413a487c1c095fa31/php_mailparse_rfc822.re#L355 since `i == toks->ntokens`.

**Comment by remicollet** [comment]
@cmb69 as your patch fix some segfault, I apply it in 7df6d614e79ebdb2bf75f4913450229b55e72078

**Comment by eXorus** [comment]
Could we release the version for PHP 7.3, I will try to run my test on php-mime-mail-parser ?

**Comment by eXorus** [comment]
Hi everyone,

Thanks for your work @remicollet  @cmb69 @Jan-E @rlerdorf 
I don't see any release for PHP 7.3, why ?

**Comment by zbenc** [comment]
Thank you all for working on this.
Any chance though, we will see this released before PHP 7.4? (Please pardon the sarcasm)

**Comment by rlerdorf** [comment]
Just pull from the latest commit. Most fringe pecl extensions don't get Github releases. If you are building your own rpms, here is a .spec you can use:

```
%global commit 4f97883fb725641a0ac7b946fcfd071605510bd0

%global shortcommit %(c=%{commit}; echo ${c:0:7})
%global libdir %{_libdir}/php/modules7
%global inidir %{_sysconfdir}/php7/php.d
%global modname mailparse

Name:		php73-pecl-%{modname}
Version:	3.0.3dev
Release:	1.%{shortcommit}.%{?dist}
Summary:	PHP7 %{modname} module

License:  PHP
URL:      https://github.com/php/pecl-mail-mailparse
Source0:  https://github.com/php/pecl-mail-mailparse/archive/%{commit}.tar.gz
Source1:  %{modname}.ini
BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)

BuildRequires: php73-devel >= 7.3.0
Requires: php73 >= 7.3.0

%description
PHP %{modname} pecl module

%prep
%setup -n pecl-mail-mailparse-%{commit}

%build
%{_bindir}/phpize
./configure 
make %{?_smp_mflags}

%install
rm -rf $RPM_BUILD_ROOT
install -m 755 -d $RPM_BUILD_ROOT%{libdir}
install -m 755 -d $RPM_BUILD_ROOT%{inidir}
install -m 644 modules/%{modname}.so $RPM_BUILD_ROOT%{libdir}
install -m 644 %{SOURCE1} $RPM_BUILD_ROOT%{inidir}

%clean
rm -rf $RPM_BUILD_ROOT

%files
%defattr(-,root,root,-)
%{libdir}/%{modname}.so
%{inidir}/%{modname}.ini
%doc

%check
make test

%changelog
```

**Comment by nikic** [comment]
@rlerdorf Is it possible to give @remicollet rights to publish all pecl packages?

**Comment by rlerdorf** [comment]
Sure, just add him.

**Comment by eXorus** [comment]
@remicollet  When do you think it will be possible to publish the new release on PECL?

Thanks for your help.

**Comment by eXorus** [comment]
Sorry to insist but why do we merge this PR on master without releasing it ? I don't understand.

**Comment by remicollet** [comment]
@eXorus simply because the current **lead** is inactive, and this is its work, or giving some other dev this role.

In short: dead and unmaintained extension.

**Comment by cmb69** [comment]
@eXorus Consider to apply as maintainer for mailparse via <https://pecl.php.net/account-request.php>.

**Comment by globules-io** [comment]
@eXorus Did you find a solution? Looking for the 7.3 DLL here...

**Comment by Jan-E** [comment]
Try the DLL's in the zips here:
https://www.apachelounge.com/viewtopic.php?t=6359

They are built from git-headof this repository.

**Comment by eXorus** [comment]
Sorry @remicollet @cmb69  @globules-io @Jan-E 
but I didn't receive any response from my request to become a maintainer...

mailparse is dead

**Comment by Jan-E** [comment]
On the mailinglist there was a question by @remicollet 

Remi Collet (Fri, 8 Feb 2019 14:06:37 +0100):
>Le 08/02/2019 à 13:31, Vincent Dauce a écrit : 
> > We always have some difficulties to publish a new version of mailparse and I would like to help.
> > The last release of Mailparse is not compatible with PHP 7.2 but the master branch is compatible with PHP 7.3 and we are waiting for a release on pecl as discussed here: 
> 
>But do you have the skill to maintain this extension ?
>
>If this is only about a "QA release", I can probably roll one.
>
>But this extension really need a new maintainer / lead.
>
>Remi

**Comment by globules-io** [comment]
@Jan-E Thanks, life saver

**Comment by eXorus** [comment]
@Jan-E @remicollet I didn't see the response in php.pecl.dev.

I tried to reply but I can't, it's the first time I use newgroup
```
<pecl-dev@lists.php.net>:
Sorry, only subscribers may post. If you are a subscriber, please forward this message to pecl-dev-owner@lists.php.net to get your new address included.
```

So this is my response ;)

> I don't know the skill to maintain an PHP extension but I can learn.
> 
> Since 2013, I'm the mainainer of the lib php-mime-mail-parser (more than
> 600k downloads) that is a wrapper of mailparse so I have a lof of interest.
> I have a master degree in computer sciences so I can code in any language.
> And I'm working as a Quality Assurance Director in a startup named
> OpenClassrooms so I know exactly what is the quality expected for a
> software.

**Comment by Jan-E** [comment]
As PHP extension maintainer you should at least follow the pecl-dev list. Send a message to pecl-dev-help@lists.php.net to get gelp on subcribing.

**Comment by eXorus** [comment]
I just installed PHP7.3 and build the master branch of mailparse to test https://github.com/php-mime-mail-parser/php-mime-mail-parser.

It works very well.

```
➜  php-mime-mail-parser git:(master) ✗ vendor/bin/phpunit
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

...............................................................  63 / 151 ( 41%)
............................................................... 126 / 151 ( 83%)
.........................                                       151 / 151 (100%)

Time: 604 ms, Memory: 27.12MB

OK (151 tests, 1863 assertions)
```

```
➜  php-mime-mail-parser git:(master) ✗ /usr/bin/php7.2 vendor/bin/phpunit 
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

...............................................................  63 / 151 ( 41%)
............................................................... 126 / 151 ( 83%)
.........................                                       151 / 151 (100%)

Time: 634 ms, Memory: 27.12MB

OK (151 tests, 1863 assertions)
```

I have now a PECL account but I don't have any access to publish the next version of mailparse.

**Comment by eXorus** [comment]
For all of you that are waiting the new mailparse version for PHP 7.3, check here:
https://pecl.php.net/package/mailparse/3.0.3

Thanks @remicollet

**Comment by sanmai** [comment]
I'm still having persistent segfaults with the latest version. How should I debug these?

**Comment by eXorus** [comment]
@sanmai did you try the installation from here?
https://github.com/php-mime-mail-parser/php-mime-mail-parser

Depends of your OS most of the time.

**Comment by sanmai** [comment]
Yeah, that's what I'm using. And I'm an idiot, I wasn't using the most recent version.

```
Preparing to unpack .../php-mailparse_3.0.3~dev20181093+2.1.7~dev20160128-1_amd64.deb ...
Unpacking php-mailparse (3.0.3~dev20181093+2.1.7~dev20160128-1) ...
Setting up php-mailparse (3.0.3~dev20181093+2.1.7~dev20160128-1) ...
```
Let's hope my problems will be solved with the newest version.

**Pull request closed without merging**